### PR TITLE
Update SNP Verifier with report and init claims

### PR DESCRIPTION
**Pull request by fitzthum**
Also, move the SNP testing data into the test_data directory.

@mkulke This seems like it's going to have somewhat incorrect implications for the AZ Verifier (since we don't want to use the init data from the report for that). Not sure what the best way to reconcile is. 

@Xynnn007 It seems like we don't specify what format the report_data and init_data should be returned in the claims map. I am assuming we want base64.

I am planning to also add to the parsed claims doc once https://github.com/confidential-containers/kbs/pull/248 is merged.

### CODE REVIEW

**Comment by Xynnn007** [comment]
Now SGX and TDX are using hex as the claims. Maybe we should have a standard for all platforms? All using base64 or hex? It would help the consumer without worrying about the encoding of the claims.

**Comment by Xynnn007** [comment]
btw #248 is mergwd

**Comment by fitzthum** [comment]
> Now SGX and TDX are using hex as the claims. Maybe we should have a standard for all platforms? All using base64 or hex? It would help the consumer without worrying about the encoding of the claims.

Yeah I think it has to be standardized or we aren't going to be able to move the checks out of individual verifiers.

**Comment by mkulke** [comment]
If the claims are generic for each TEE, should we add them individually in the TEE-specific parts of a the verification logic? can we just add them after the verification has passed?

https://github.com/confidential-containers/kbs/blob/eebaa977e0a23febb5318994cb0e5ba9daa7f8d5/attestation-service/attestation-service/src/lib.rs#L102-L116

(IMO, it probably would be cleaner to have them added in the `verifier` crate, so there would be `evidence -> claim_set` logic, be but we're already messing with the claim (flattening) in the `attestation-service` snippet above, so that might be ok)

**Comment by mkulke** [comment]
I recommend decoupling test cleanups into a discrete PR. We should really attempt to reduce the number of drive-by fixes. Non-controversial but necessary quality improvements could be merged quickly, and we can focus our review-time on parts which require more scrutiny. untangling unrelated parts from change-sets is a cognitive overhead.

**Comment by mkulke** [comment]
@fitzthum 

> As of today we could add the expected init_data and report_data values to the claims map after the evidence has been verified, but I think the next step is to remove verification of those fields from the verifiers.

For az-{snp,tdx}-vtpm TEEs we would have to make changes to support this, at the moment nonce/report_data verification is coupled with signature validation in an encapsulated library fn, like this:

image

but in principle the verification can be split, I think. What would be the projected timeframe for such a change?

> Instead, the verifier will just return the values as part of the claims map and the AS will make sure they are correct. I think this approach will make the interface a bit cleaner and help decouple verification from evaluation. One downside is that it will expose the report_data which is not necessary and might be confusing.

Is there a reason why report_data should be part of a claims map? It's a property of a concrete RCAR exchange and hence not useful for a consumer of a claims map, is it?

Evidence extraction logic for report_data would still need to be implemented individually for each TEE, even if the verification is moved to upper layers. I would still recommend to have this close to the verifier implementation, maybe we can change the trait signatures, so a given TEE's evidence needs to implement `report_data()`:

```Rust
trait ExtractReportData {
    fn report_data(&self) -> &[u8];
}

type ClaimsMap = (); // dummy

trait Verifier {
    type Evidence: ExtractReportData;
    
    fn evaluate(&self, evidence: &Self::Evidence) -> Result<ClaimsMap>;
}


struct MyTee;
type MyEvidence = [u8;64]; // dummy

impl ExtractReportData for MyEvidence {
    fn report_data(&self) -> &[u8] {
        todo!();
    }
}

impl Verifier for MyTee {
    type Evidence = MyEvidence;
    
    fn evaluate(&self, evidence: &MyEvidence) -> Result<ClaimsMap> {
       todo!(); 
    }
}
```

> I am assuming [init-data] is consumed by azure to provision the vTPM.

I'm not sure I understand. Init-data is a CoCo concept, how would that be related to provisioning of a vTPM?

**Comment by fitzthum** [comment]
> Can you add this change to bottom of the az-snp-vtpm verifier logic, so we unset the respective keys from the json map?

If we drop both of these the AS isn't going to be able to verify that the report_data is correct. I guess we could drop the init_data from the AZ SNP claims. The check in the AS for the init_data can maybe be optional?

**Comment by mkulke** [comment]
> > Can you add this change to bottom of the az-snp-vtpm verifier logic, so we unset the respective keys from the json map?
> 
> If we drop both of these the AS isn't going to be able to verify that the report_data is correct. I guess we could drop the init_data from the AZ SNP claims. The check in the AS for the init_data can maybe be optional?

re report_data:

as long as the signature to `verifier::evaluate()` doesn't change and we still receive `expected_report_data` as argument, we can just pass it to the claim after verification and verify it again in upper layers:

```
async fn evaluate(
      &self,
      evidence: &[u8],
====> expected_report_data: &ReportData,
      expected_init_data_hash: &InitDataHash,
)
```

If the signature changes, and we don't receive `expected_report_data` any more, then there will be breakage, b/c atm we cannot extract the nonce from the evidence.

re init_data:

at the moment it's still not entirely clear to me how we want to curate init_data in attestation-agent on a peerpod. If we can ignore it for the time being until we know more, that would be fine w/ me.

**Comment by fitzthum** [comment]
> If the signature changes, and we don't receive expected_report_data any more, then there will be breakage, b/c atm we cannot extract the nonce from the evidence.

If there is no way to extract the report data from the evidence, then https://github.com/confidential-containers/kbs/issues/228 might not really be feasible.

**Comment by mkulke** [comment]
> If there is no way to extract the report data from the evidence, then #228 might not really be feasible.

Ah, I just meant it won‘t work automagically ootb, we need to make a crate release + bump a dep in this repo before we can use the respective code to do that. in principle #228 shouldn‘t be blocked, just needs prepwork

**Comment by fitzthum** [comment]
> Ah, I just meant it won‘t work automagically ootb, we need to make a crate release + bump a dep in this repo before we can use the respective code to do that. in principle https://github.com/confidential-containers/kbs/issues/228 shouldn‘t be blocked, just needs prepwork

Ok, in that case let me add the code you suggested and once we are ready we can tweak things to work better for #228 

btw I think in the long term we can just have the AS remove the report_data from the claims map before it does the policy evaluation.

**Comment by fitzthum** [comment]
Ok, I forget what was happening with this PR. I think we ran into some confusion regarding how to handle `init_data` and `report_data`. I have dropped the commit that adds these to the SNP claims map. Now this PR just moves the test files and adds a description of the SNP claims to the doc. PTAL @mkulke @Xynnn007

File: attestation-service/verifier/src/snp/mod.rs [hunk 1]
```diff
@@ -258,6 +258,11 @@ mod tests {
     use super::*;
     use openssl::nid::Nid;
 
+    const VCEK: &[u8; 1360] = include_bytes!("../../test_data/snp/test-vcek.der");
```
Comment (Xynnn007) [inline attestation-service/verifier/src/snp/mod.rs:261]: Might not related to the PR. But if we can have an agreement that the binary values should be all showed as hex or base64? see Line 250 in this file (I cannot add comment to line out of this PR)

File: attestation-service/docs/parsed_claims.md [hunk 2]
```diff
@@ -92,3 +92,27 @@ The claim inherit the fields from the SEV-SNP claim with and additional `tpm` hi
...
+- `policy_abi_major`: Minimum ABI major version allowed for guest
+- `policy_abi_minor`: Minimum ABI minor version allowed for guest
+- `policy_debug_allowed`: Whether SNP debug features are allowed for guest
+- `policy_migrate_ma`: Whether migration agent can be connected to guest
+- `policy_single_socket`: Whether guest can be activated only on one socket
+- `policy_smt_allowed`: Whether guest can run on a system with SMT enabled
+- `reported_tcb_bootloader`: Reported SVN of ASP bootloader
+- `reported_tcb_microcode`: Reported microcode version
+- `reported_tcb_snp`: Reported SVN of SNP Firmware
+- `reported_tcb_tee`: Reported SVN of ASP OS
```
Comment (Xynnn007) [inline attestation-service/docs/parsed_claims.md:110]: all claims above would be add a prefix `snp.` due to the function https://github.com/confidential-containers/trustee/blob/main/attestation-service/attestation-service/src/utils.rs#L39

**Review by Xynnn007 — COMMENTED** [review]
Thanks for this. The PR brings me to old times.. ; P

File: attestation-service/verifier/src/snp/mod.rs (hunk 1 above)
Comment (fitzthum) [reply attestation-service/verifier/src/snp/mod.rs:261]: Yeah let's take this in a separate issue but I think it is probably a good idea.

Comment (Xynnn007) [reply attestation-service/verifier/src/snp/mod.rs:261]: Ok. See https://github.com/confidential-containers/trustee/issues/372. And I think other fields of SNP should also do the same, such as https://github.com/confidential-containers/trustee/blob/main/attestation-service/verifier/src/snp/mod.rs#L240

**Review by mkulke — APPROVED** [review]
lgtm

**Comment by fitzthum** [comment]
Rebased to pick up the vlek change. I think the comments are accounted for so I merged before I forget about this PR again.

**Pull request merged**